content/md/posts/2023-07-08-testing.md

```diff
@@ -0,0 +1,150 @@
+          name (rf/subscribe [::subs/name])]
+      (t/are [x y] (= x y)
+        :registered @user
+        "han@skywalker.com" @name)))))
+```
+
+First, I initialize my routes, setup the database, and connect to firebase. After that, I login and wait for the user's information to be fetched. I then check whether the name of the user matches the credentials entered earlier and if the user is noted as registered.
+
+## Acceptance testing
+
+Acceptance testing is a bit different from the first two tests, as it tests things from the perspective of the user. You think of what the user would need to do to in order to do something, such as logging in, and you want to test whether what they see matches what should be there.
+
+For example, in my application when the user logs in they are taken to their dashboard and shown a welcome message. I want to see that when they reach the dashboard, they see the welcome message. To make this test, we need to set a few things up.
+
+Setup the directory for acceptance testing as directed in the post "Setting up the project part 2". After, create a file in test/acceptance called admin_login.clj. Here, we are going to put the test that we described above.
+
+Change the namespace so it appears like this. 
+```clojure
+(ns acceptance.admin-login-test
+  (:require  [clojure.test :as t]
+             [etaoin.api :as e]
+             [etaoin.keys :as k]
+             [acceptance.core :as c :refer [driver]])
+  (:import com.google.firebase.cloud.FirestoreClient))
+```
+
+Then, add the following code to the file.
+
+```clojure
+(def default-message
+  "Hello (Unknown). We're glad to see you.")
+
+(def admin-login-message
+  "Hello han@skywalker.com. Welcome to your dashboard. To make a new document, click the button to your right. To share and edit your existing documents, look at your documents below")
+
+(t/use-fixtures :once c/init-firebase) ;; Sets up firebase once for each test
+(t/use-fixtures :each c/with-screenshot) ;; Takes a screenshot at the end of every test
+
+(t/deftest message-test
+  (t/testing "When the admin user exists"
+    (doto driver
+      (c/home->dashboard)
+      (e/wait-has-text-everywhere admin-login-message))
+    (let [actual (e/get-element-text driver {:data-testid :user})]
+      (e/screenshot driver "screenshots/message-test-when-the-admin-exists.png")
+      (t/is (= admin-login-message actual)))))
+```
+
+In this test, I login and head to the dashboard, and then see if the welcome message has appeared. If it does not appear, an error is thrown. If it is found, I check to see if the welcome message matches the admin login message.
+
+
+
+## Future posts
+
+For future posts about testing, I will talk about new situations I had to deal with in regards to testing, or my thoughts on how tests could be made easier to understand. I will also include more posts which focus on a specific type of test, such as unit/acceptance/integration, and expand on my experience with that type of test so far.
```

Comment:
NIT: I believe a bit more succinct could be better

```
On my next posts I will describe in more detail the setup and gotchas I found with unit, integration and acceptance testing.
```

Reply:
Ok.